src/vs/workbench/contrib/chat/browser/chatContentParts/toolInvocationParts/chatTerminalToolProgressPart.ts

```diff
@@ -754,168 +927,321 @@ class ChatTerminalToolOutputSection extends Disposable {
-				inlineTerminal.style.setProperty('color', theme.foreground || 'inherit');
-			}
+		if (this.isExpanded) {
+			this._scheduleOutputRelayout();
 		}
+		this._updateTerminalVisibility();
 
-		this._outputBody.replaceChildren(content);
-		this._outputContent = content;
-		if (!this._outputScrollbar) {
-			this._outputScrollbar = this._register(new DomScrollableElement(this._outputBody, {
-				vertical: ScrollbarVisibility.Auto,
-				horizontal: ScrollbarVisibility.Auto,
-				handleMouseWheel: true
-			}));
-			const scrollableDomNode = this._outputScrollbar.getDomNode();
-			scrollableDomNode.tabIndex = 0;
-			scrollableDomNode.style.maxHeight = `${MAX_TERMINAL_OUTPUT_PREVIEW_HEIGHT}px`;
-			this._container.appendChild(scrollableDomNode);
-			this._ensureOutputResizeObserver();
-			this._outputContent = undefined;
-			this._renderedOutputHeight = undefined;
-		} else {
-			this._ensureOutputResizeObserver();
+		if (wasTrimmed || this._bufferedLineCount >= maxLines) {
+			this._lastOutputTruncated = true;
+			this._isStreaming = false;
+			storedOutput.truncated = true;
+			const combined = this._streamBuffer.join('');
+			storedOutput.text = combined;
+			this._setStatusMessages();
+			this._updateTerminalVisibility();
+			return true;
 		}
-		this.updateAriaLabel();
-		return true;
+		storedOutput.truncated = false;
+		this._updateTerminalVisibility();
+		const combined = this._streamBuffer.join('');
+		storedOutput.text = combined;
+		storedOutput.truncated = this._lastOutputTruncated;
+		return false;
 	}
 
-	private async _collectOutput(terminalInstance: ITerminalInstance | undefined): Promise<{ text: string; truncated: boolean } | undefined> {
-		const commandDetection = terminalInstance?.capabilities.get(TerminalCapability.CommandDetection);
-		const commands = commandDetection?.commands;
-		const xterm = await terminalInstance?.xtermReadyPromise;
-		if (!commands || commands.length === 0 || !terminalInstance || !xterm) {
+	public applyStreamingSnapshot(snapshot: string): void {
+		// Applies a serialized VT snapshot captured from the command markers. We try to diff the
+		// new snapshot against the previously seen content to avoid costly full replays.
+		if (this._lastRawSnapshot === snapshot) {
+			return;
+		}
+		if (!this._lastRawSnapshot) {
+			this._replaceStreamingDataFromRaw(snapshot);
+			this._lastRawSnapshot = snapshot;
+			return;
+		}
+		if (snapshot.startsWith(this._lastRawSnapshot)) {
+			const appended = snapshot.slice(this._lastRawSnapshot.length);
+			if (appended.length) {
+				this.appendStreamingData(appended);
+			}
+			this._lastRawSnapshot = snapshot;
 			return;
 		}
-		const commandId = this._terminalData.terminalCommandId ?? this._getStoredCommandId();
-		if (!commandId) {
+		if (this._lastRawSnapshot.startsWith(snapshot)) {
+			this._replaceStreamingDataFromRaw(snapshot);
+			this._lastRawSnapshot = snapshot;
 			return;
 		}
```

Comment:
This approach will be slow when the terminal commands are large (`largeString.startsWith(largeString2)` needs to iterate `largeString2.length` chars). Additionally I think when the command starts to get trimmed it'll fail entirely and reprint everything since the beginning will be slow.